core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1805,10 +1805,10 @@ class Log(@volatile var dir: File,
       // delete the old files
       for (seg <- sortedOldSegments) {
         // remove the index entry
-        if (seg.baseOffset != sortedNewSegments.head.baseOffset)
-          segments.remove(seg.baseOffset)
-        // delete segment
-        asyncDeleteSegment(seg)
+        if (seg.baseOffset != sortedNewSegments.head.baseOffset) {
+          if (segments.remove(seg.baseOffset) != null)
+            asyncDeleteSegment(seg)
```

Comment:
The way this is implemented now, we will prevent the segment with base offset equal to `sortedNewSegments.head.baseOffset` from being deleted.

Reply:
Good catch. My bad. I have updated the patch to fix the issue.

The latest patch keeps the original order of disk operation in `replaceSegments()`, which first changes suffix to `.swap` for new segments, change suffix to `.delete` for old segments, and change suffix for new segments again. So it should handle the server crash in the same manner as before this patch.

The latest patch slightly changed the order in which `replaceSegments()` updates the segment index. Previously we first add new segment to the index, and remove old segment from the index if base offset of the old segment meets some criteria. In the this patch `replaceSegments()` will first remove the old segment from the index, and then add new segment to the index. This solution seems more intuitive and a bit cleaner as it does not require the if statement `seg.baseOffset != sortedNewSegments.head.baseOffset`.